File: next/OpenCode/OpencodeHooks.fs
namespace Wanxiangshu.Next.OpenCode

open System
open System.Threading
open Fable.Core
open Fable.Core.JsInterop
open Wanxiangshu.Next.Kernel
open Wanxiangshu.Next.Kernel.Identity
open Wanxiangshu.Next.Kernel.Fact
open Wanxiangshu.Next.Journal
open Wanxiangshu.Next.Session

module OpencodeHooks =

    let private processUserMessage
        (gateway: Gateway)
        (drivers: SessionDrivers)
        (inboxMap: System.Collections.Generic.Dictionary<SessionId, ISessionInbox>)
        (sessionId: SessionId)
        (userMsg: OpencodeUserMessage)
        =
        let key: SessionDriversKey =
            { RuntimeId = gateway.RuntimeId
              SessionId = sessionId }

        let origin = MessageOriginDecoder.decodeUserMessageOrigin userMsg

        let inbox =
            match inboxMap.TryGetValue(sessionId) with
            | true, ib -> ib
            | false, _ ->
                let ib = FifoInbox(1000) :> ISessionInbox
                inboxMap.[sessionId] <- ib
                ib

        let text =
            userMsg.parts
            |> List.choose (fun p -> if isNull p?text then None else Some(unbox<string> p?text))
            |> String.concat "\n"

        match origin with
        | Human turnId ->
            drivers.BumpLocalEpochOnHuman key |> ignore
            inbox.TryPost(HumanMessageEvent(turnId, text)) |> ignore
            let cts = new CancellationTokenSource()
            drivers.Activate(key, cts) |> ignore
        | PluginGenerated promptKeyRef -> inbox.TryPost(PluginEvent(PromptKeyRef.value promptKeyRef, text)) |> ignore
        | HostInternal -> ()

    let handleChatMessage
        (gateway: Gateway)
        (drivers: SessionDrivers)
        (inboxMap: System.Collections.Generic.Dictionary<SessionId, ISessionInbox>)
        (input: OpencodeHookInput)
        (outputObj: obj)
        =
        if not (isNull outputObj) && not (isNull outputObj?message) then
            let msg = outputObj?message
            let role = unbox<string> msg?role

            if role = "user" then
                let sessionId = SessionId.create input.sessionID

                let rawModel =
                    match input.model with
                    | Some m -> Some m
                    | None ->
                        if not (isNull msg?model) then
                            let mObj = msg?model

                            Some
                                { providerID = unbox<string> mObj?providerID
                                  modelID = unbox<string> mObj?modelID
                                  variant =
                                    if isNull mObj?variant then
                                        None
                                    else
                                        Some(unbox<string> mObj?variant) }
                        else
                            None

                let rawParts = if isNull msg?parts then [] else unbox<obj list> msg?parts

                let rawAgent =
                    match input.agent with
                    | Some a -> Some a
                    | None ->
                        if isNull msg?agent then
                            None
                        else
                            Some(unbox<string> msg?agent)

                let userMsg: OpencodeUserMessage =
                    { id = unbox<string> msg?id
                      role = role
                      sessionID = input.sessionID
                      agent = rawAgent
                      model = rawModel
                      parts = rawParts }

                processUserMessage gateway drivers inboxMap sessionId userMsg

    let handleToolExecuteBefore (input: OpencodeToolExecuteInput) (output: OpencodeToolExecuteOutput) =
        let argsObj = output.args

        if not (isNull argsObj) then
            if not (isNull argsObj?warn_tdd) then
                argsObj?warn_tdd <- null

            if not (isNull argsObj?warn_reuse) then
                argsObj?warn_reuse <- null

            if not (isNull argsObj?warn_context) then
                argsObj?warn_context <- null

    let handleEvent
        (gateway: Gateway)
        (inboxMap: System.Collections.Generic.Dictionary<SessionId, ISessionInbox>)
        (eventObj: obj)
        =
        if not (isNull eventObj) then
            let eventType = unbox<string> eventObj?``type``
            let properties = eventObj?properties

            if not (isNull properties) && not (isNull properties?sessionID) then
                let sessionId = SessionId.create (unbox<string> properties?sessionID)

                match inboxMap.TryGetValue(sessionId) with
                | true, inbox ->
                    match eventType with
                    | "session.idle" -> inbox.TryPost(LifecycleEvent "session.idle") |> ignore
                    | "session.status" ->
                        let statusType = unbox<string> properties?status?``type``
                        inbox.TryPost(LifecycleEvent($"session.status:{statusType}")) |> ignore
                    | "message.updated" ->
                        let msgObj = properties?info

                        if not (isNull msgObj) then
                            let role = unbox<string> msgObj?role

                            if role = "assistant" then
                                let msgId = MessageId.create (unbox<string> msgObj?id)

                                let parentId =
                                    if isNull msgObj?parentID then
                                        MessageId.create ""
                                    else
                                        MessageId.create (unbox<string> msgObj?parentID)

                                let isError = not (isNull msgObj?error)

                                let outcome =
                                    if isError then
                                        PromptOutcome.FatalFailure "assistant-error"
                                    else
                                        PromptOutcome.Delivered msgId

                                inbox.TryPost(AssistantTerminalEvent(parentId, msgId, outcome)) |> ignore
                    | _ -> ()
                | false, _ -> ()

File: .wanxiangshu/runtimes/OpenCode/OpencodeHooks.js

import { SessionDrivers__Activate_3FD309E6, SessionDrivers__BumpLocalEpochOnHuman_Z6488BD67, SessionDriversKey } from "../Session/Driver.js";
import { decodeUserMessageOrigin } from "./MessageOriginDecoder.js";
import { defaultOf } from "../fable_modules/fable-library-js.5.6.0/Util.js";
import { tryGetValue } from "../fable_modules/fable-library-js.5.6.0/MapUtil.js";
import { FSharpRef } from "../fable_modules/fable-library-js.5.6.0/Types.js";
import { SessionInboxEvent, FifoInbox_$ctor_Z524259A4 } from "../Session/Inbox.js";
import { concat, join } from "../fable_modules/fable-library-js.5.6.0/String.js";
import { empty, choose } from "../fable_modules/fable-library-js.5.6.0/List.js";
import { Operators_IsNull } from "../fable_modules/fable-library-js.5.6.0/FSharp.Core.js";
import { MessageIdModule_create, SessionIdModule_create, PromptKeyRefModule_value } from "../Kernel/Identity.js";
import { createCancellationToken } from "../fable_modules/fable-library-js.5.6.0/Async.js";
import { OpencodeUserMessage, OpencodeModel } from "./OpencodeTypes.js";
import { PromptOutcome } from "../Kernel/Fact.js";

function processUserMessage(gateway, drivers, inboxMap, sessionId, userMsg) {
    const key = new SessionDriversKey(gateway.RuntimeId, sessionId);
    const origin = decodeUserMessageOrigin(userMsg);
    let inbox;
    let matchValue;
    let outArg = defaultOf();
    matchValue = [tryGetValue(inboxMap, sessionId, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (matchValue[0]) {
        inbox = matchValue[1];
    }
    else {
        const ib_1 = FifoInbox_$ctor_Z524259A4(1000);
        inboxMap.set(sessionId, ib_1);
        inbox = ib_1;
    }
    const text = join("\n", choose((p) => {
        if (Operators_IsNull(p.text)) {
            return undefined;
        }
        else {
            return p.text;
        }
    }, userMsg.parts));
    switch (origin.tag) {
        case 1: {
            inbox.TryPost(new SessionInboxEvent(1, [PromptKeyRefModule_value(origin.fields[0]), text]));
            break;
        }
        case 2: {
            break;
        }
        default: {
            SessionDrivers__BumpLocalEpochOnHuman_Z6488BD67(drivers, key);
            inbox.TryPost(new SessionInboxEvent(0, [origin.fields[0], text]));
            SessionDrivers__Activate_3FD309E6(drivers, key, createCancellationToken());
        }
    }
}

export function handleChatMessage(gateway, drivers, inboxMap, input, outputObj) {
    if (!Operators_IsNull(outputObj) && !Operators_IsNull(outputObj.message)) {
        const msg = outputObj.message;
        const role = msg.role;
        if (role === "user") {
            const sessionId = SessionIdModule_create(input.sessionID);
            let rawModel;
            const matchValue = input.model;
            if (matchValue == null) {
                if (!Operators_IsNull(msg.model)) {
                    const mObj = msg.model;
                    rawModel = (new OpencodeModel(mObj.providerID, mObj.modelID, Operators_IsNull(mObj.variant) ? undefined : mObj.variant));
                }
                else {
                    rawModel = undefined;
                }
            }
            else {
                rawModel = matchValue;
            }
            const rawParts = Operators_IsNull(msg.parts) ? empty() : msg.parts;
            let rawAgent;
            const matchValue_1 = input.agent;
            rawAgent = ((matchValue_1 == null) ? (Operators_IsNull(msg.agent) ? undefined : msg.agent) : matchValue_1);
            processUserMessage(gateway, drivers, inboxMap, sessionId, new OpencodeUserMessage(msg.id, role, input.sessionID, rawAgent, rawModel, rawParts));
        }
    }
}

export function handleToolExecuteBefore(input, output) {
    const argsObj = output.args;
    if (!Operators_IsNull(argsObj)) {
        delete argsObj["warn_tdd"];
        delete argsObj["warn_reuse"];
        delete argsObj["warn_context"];
    }
}

export function handleEvent(gateway, inboxMap, eventObj) {
    if (!Operators_IsNull(eventObj)) {
        const eventType = eventObj.type;
        const properties = eventObj.properties;
        if (!Operators_IsNull(properties) && !Operators_IsNull(properties.sessionID)) {
            const sessionId = SessionIdModule_create(properties.sessionID);
            let matchValue;
            let outArg = defaultOf();
            matchValue = [tryGetValue(inboxMap, sessionId, new FSharpRef(() => outArg, (v) => {
                outArg = v;
            })), outArg];
            if (matchValue[0]) {
                const inbox = matchValue[1];
                switch (eventType) {
                    case "session.idle": {
                        inbox.TryPost(new SessionInboxEvent(6, ["session.idle"]));
                        break;
                    }
                    case "session.status": {
                        const statusType = properties.status.type;
                        inbox.TryPost(new SessionInboxEvent(6, [concat("session.status:", statusType)]));
                        break;
                    }
                    case "message.updated": {
                        const msgObj = properties.info;
                        if (!Operators_IsNull(msgObj)) {
                            if (msgObj.role === "assistant") {
                                const msgId = MessageIdModule_create(msgObj.id);
                                const parentId = Operators_IsNull(msgObj.parentID) ? MessageIdModule_create("") : MessageIdModule_create(msgObj.parentID);
                                const outcome = !Operators_IsNull(msgObj.error) ? (new PromptOutcome(3, ["assistant-error"])) : (new PromptOutcome(0, [msgId]));
                                inbox.TryPost(new SessionInboxEvent(2, [parentId, msgId, outcome]));
                            }
                        }
                        break;
                    }
                    default:
                        undefined;
                }
            }
        }
    }
}

